src/server/query.ml

```diff
@@ -28,8 +28,11 @@ let all_queries request =
   |> snd
   |> Formats.from_form_urlencoded
 
-let query request name =
-  List.assoc_opt name (all_queries request)
+let query request ?(filter_empty = false) name =
+  match (List.assoc_opt name (all_queries request), filter_empty) with
```

Comment:
I recommend rewriting this as

```ocaml
match List.assoc_opt name (all_queries request) with
| None -> None
| Some v as result ->
  if filter_empty then
    result
  else
    if String.trim v = "" then
      None
    else
      result
```

This avoids unconditionally allocating `Some` options every time there is a `Some v` match. Also, it avoids mixing in the `filter_empty` check into all branches, when it is only relevant for the `Some v` branch.

Later, if actually optimizing, we may want to replace `String.trim` by a scan of the string that does not allocate an intermediate string potentially almost as long as `v`, only for us to compare it with `""`.

Reply:
> This avoids unconditionally allocating Some options every time there is a Some v match. Also, it avoids mixing in the filter_empty check into all branches, when it is only relevant for the Some v branch.

Oh, that makes sense, thanks for the explanation 😊 I think the conditions should be inverted in your example. It should be:

```ocaml
match List.assoc_opt name (all_queries request) with
| None -> None
| Some v as result ->
  if filter_empty then
    if String.trim v = "" then
      None
    else
      result
  else
    result
```

Which I've simplified to:

```ocaml
match List.assoc_opt name (all_queries request) with
| None -> None
| Some v as result ->
  if filter_empty && String.trim v = "" then
    None
  else
    result
```

I hope that makes sense.

Reply:
> Later, if actually optimizing, we may want to replace String.trim by a scan of the string that does not allocate an intermediate string potentially almost as long as v, only for us to compare it with "".

Interesting, would something like the following be reasonable ?

```ocaml
(* taken from https://github.com/ocaml/ocaml/blob/trunk/stdlib/string.ml#L97 *)
let is_space = function
  | ' ' | '\012' | '\n' | '\r' | '\t' -> true
  | _ -> false
  
let is_empty s =
  let len = String.length s in
  if len = 0 then
    true
  else
    let rec loop len i =
      if i < len && is_space s.[i] then
        loop len (i + 1)
      else
        false
    in
    loop len 0
```

Or without the inner `loop` function:

```ocaml
let is_empty s =
  let len = String.length s in
  if len = 0 then
    true
  else
    let i = ref 0 in
    let contains_only_spaces = ref true in
    while !contains_only_spaces && !i < len do
      if is_space s.[!i] then
        i := !i + 1
      else
        contains_only_spaces := false
    done;
    !contains_only_spaces
```